### ISSUE TITLE
Functions returning type of one of field descriptors in dataclass transform aren't allowed

### ISSUE THREAD

**Issue by patrick91** [microsoft/pyright#2699]
Hi there, I am using the dataclass transform for my library and it works quite well, but I found a case where it is not working 
as I expected.

I have some users creating a custom `field` function that returns an instance of the field descriptor, like so:

```python
@strawberry.type
class X:
    name: str = my_custom_field()  # this class strawberry.field under the hood
```

I've made a minimal reproduction case:

### Minimal reproduction case:

```python
from typing import *

_T = TypeVar("_T")

def __dataclass_transform__(
    *,
    eq_default: bool = True,
    order_default: bool = False,
    kw_only_default: bool = False,
    field_descriptors: Tuple[Union[type, Callable[..., Any]], ...] = (()),
) -> Callable[[_T], _T]:
    return lambda a: a

class ModelField:
    def __init__(self, *, init: bool = True, default: Optional[Any] = None) -> None:
        ...

def model_field(
    *, init: bool = True, default: Optional[Any] = None, alias: Optional[str] = None
) -> Any:
    ...

@__dataclass_transform__(
    kw_only_default=True,
    field_descriptors=(ModelField, model_field),
)
class ModelMeta(type):
    def __init_subclass__(
        cls,
    ) -> None:
        ...

class ModelBase(metaclass=ModelMeta):
    ...

def custom_field() -> ModelField:
    return ModelField()

class Customer1(ModelBase, frozen=True):
    id: int = model_field()
    name: str = model_field()
    name2: str = model_field(alias="other_name", default="None")
    name3: str = custom_field()   # this is the line that complains
```

The error I get is:

```
Expression of type "ModelField" cannot be assigned to declared type "str"
  "ModelField" is incompatible with "str"
```

I'd expect this to work since `ModelField` was passed to the field descriptors :)

**Comment by erictraut** [comment]
The `custom_field` function would need to return `Any`, just like `model_field` does. The problem is that the field descriptor type disagrees with the annotation on the field, so the convention established by the stdlib dataclass library is to always annotate the field descriptor function as returning `Any`.

Also, you would need to specify `custom_field` in the `__dataclass_transform__` since it's the function that the transform needs to understand as allocating the field descriptor.

**Comment by patrick91** [comment]
> The `custom_field` function would need to return `Any`, just like `model_field` does. The problem is that the field descriptor type disagrees with the annotation on the field, so the convention established by the stdlib dataclass library is to always annotate the field descriptor function as returning `Any`.

Gotcha! that works 😊

> Also, you would need to specify `custom_field` in the `__dataclass_transform__` since it's the function that the transform needs to understand as allocating the field descriptor.

I can't do that as the `custom_field` function is defined by the user of the library, so I don't have control over that :(

**Comment by erictraut** [comment]
Could you explain the use case for `custom_field`? Where would someone use that instead of `model_field`? The stdlib `dataclass` doesn't support this, at least in any way that would be compatible with static type checking.

If `custom_field` is intended to supply some of the arguments to the `ModelField` constructor without requiring the caller to pass those same arguments to `custom_field`, that would be problematic because the type checker would have no way of knowing what those arguments were. For example, if the `custom_field` function supplied a `default=None` argument to `ModelField`, the type checker wouldn't know that it should treat that field as having a default value.

**Comment by patrick91** [comment]
> If `custom_field` is intended to supply some of the arguments to the `ModelField` constructor without requiring the caller to pass those same arguments to `custom_field`, that would be problematic because the type checker would have no way of knowing what those arguments were. For example, if the `custom_field` function supplied a `default=None` argument to `ModelField`, the type checker wouldn't know that it should treat that field as having a default value.

That's a good point and it might be problematic for our usecase

> Could you explain the use case for `custom_field`? Where would someone use that instead of `model_field`? The stdlib `dataclass` doesn't support this, at least in any way that would be compatible with static type checking.

I guess this not 100% compatible with dataclasses, but one use case is to provide additional metadata, like so:


```python
def restricted_field() -> StrawberryField:
    return strawberry.field(permission_classes=[Example])

@strawberry.type
class ABC:
    title: str
    slug: str
    description Optional[str] = restricted_field()
    abstract: Optional[str] = restricted_field()
```

**Comment by erictraut** [comment]
If the intent is to add additional metadata but not supply any arguments that matter to a static type checker (e.g. `default`, `default_factory`, `alias`), then this would still work fine. There's no need to tell `__dataclass_transform__` about `custom_field` or `restricted_field` in that case. Simply annotate the function to return `Any`, and it should type check without a problem.

If the intent is to allow `custom_field` or `restricted_field` to accept arguments that are relevant to static type checking, then this wouldn't work. For example:

```python
def restricted_field(default: object | None = None) -> Any:
    return strawberry.field(default=default, permission_classes=[Example])

@strawberry.type
class ABC:
    title: str
    slug: str
    description Optional[str] = restricted_field(default="No Description")
    abstract: Optional[str] = restricted_field(default="Abstract")
```

To support this use case, I think the standard `model_field` function would need to support `**kwargs`, allowing the caller to pass in these custom metadata fields directly.

```python
def model_field(
    *, init: bool = True, default: Optional[Any] = None, alias: Optional[str] = None, **kwargs: Any
) -> Any:
    ...

...

def restricted_field() -> StrawberryField:
    return strawberry.field(permission_classes=[Example])

@strawberry.type
class ABC:
    title: str
    slug: str
    description Optional[str] = model_field(default="No Description", permission_classes=[Example])
    abstract: Optional[str] = model_field(default="Abstract", permission_classes=[Example])
```

Does that work for you?

**Comment by patrick91** [comment]
Yup, I think that should work fine 😊 We'll probably add some guidance for our users! thank you!
